module/programs/helper.js

```diff
@@ -976,6 +977,160 @@ module.exports = class ProgramsHelper {
+              status: httpStatusCode.bad_request.status,
+              message: constants.apiResponses.PROGRAM_JOIN_FAILED
+            });      
+          }     
+          programUsersData = {
+            programId: programId,
+            userRoleInformation: data.userRoleInformation,
+            userId: userId,
+            userProfile:userProfile.data.response
+          }
+          if( appName != "" ) {
+            programUsersData['appInformation.appName'] = appName;
+          }
+          if( appVersion != "" ) {
+            programUsersData['appInformation.appVersion'] = appVersion;
+          }
+          update['$set'] = programUsersData;
+
+          //For internal calls add consent using sunbird api
+          if( callConsetAPIOnBehalfOfUser && programData[0].hasOwnProperty('requestForPIIConsent') && programData[0].requestForPIIConsent === true ){
+            if( !programData[0].rootOrganisations || !programData[0].rootOrganisations.length > 0 ) {
+              throw {
+                message: constants.apiResponses.PROGRAM_JOIN_FAILED,
+                status: httpStatusCode.bad_request.status
+              }
+            }
+            let userConsentRequestBody = {
+              "request": {
+                "consent": {
+                  "status": constants.common.REVOKED,
+                  "userId": userProfile.data.response.id,
+                  "consumerId": programData[0].rootOrganisations[0],
+                  "objectId":  programId,
+                  "objectType": constants.common.PROGRAM
+                }
+              }
+            }
+            let consentResponse = await userService.setUserConsent(userToken, userConsentRequestBody)
+            if(!consentResponse.success){
+              throw {
+                message: constants.apiResponses.PROGRAM_JOIN_FAILED,
+                status: httpStatusCode.bad_request.status
+              }
+            }
+          }
+        }
+        
+        //create or update query
+        const query = { 
+          programId: programId,
+          userId: userId
+        };
+        let joinProgram;
+        if ( data.isResource ) {
+          update['$inc'] = { noOfResourcesStarted : 1 }
+        }
+        // add record to programUsers collection
+        joinProgram = await programUsersHelper.update(query, update, { new:true, upsert:true });
+        
+        if (!joinProgram._id) {
+          throw {
+              message: constants.apiResponses.PROGRAM_JOIN_FAILED,
+              status: httpStatusCode.bad_request.status
+          }
+        }
+        joinProgram.programName = programData[0].name;
+        joinProgram.programExternalId = programData[0].externalId;
+        joinProgram.requestForPIIConsent = (programData[0].requestForPIIConsent && programData[0].requestForPIIConsent == true) ?  true : false;
+
+        //  push programUsers details to kafka
+        await kafkaProducersHelper.pushProgramUsersToKafka(joinProgram);
```

Comment:
@VISHNUDAS-tunerlabs we should push to kafka only if 2 scenario only I think - 1st when user has joined program for first time, 2nd consent has changed from no to yes or yes to no